libnvme/src/namespace.rs

```diff
@@ -0,0 +1,194 @@
+            false => Err(controller
+                .fatal_context("failed to init nvme namespace discovery")),
+        }
+    }
+
+    fn internal_step(&self) -> Result<Option<Namespace<'a>>, NvmeError> {
+        let mut nvme_ns_disc: *const nvme_ns_disc_t = std::ptr::null_mut();
+        let state =
+            unsafe { nvme_ns_discover_step(self.iter, &mut nvme_ns_disc) };
+        match state {
+            NVME_ITER_VALID => {
+                let nsid = unsafe { nvme_ns_disc_nsid(nvme_ns_disc) };
+                let mut ns: *mut nvme_ns_t = std::ptr::null_mut();
+                match unsafe {
+                    nvme_ns_init(self.controller.inner, nsid, &mut ns)
+                } {
+                    true => Ok(Some(Namespace {
+                        inner: ns,
+                        controller: self.controller,
+                    })),
+                    false => Err(self
+                        .controller
+                        .fatal_context("failed to init nvme namespace")),
+                }
+            }
+            NVME_ITER_DONE => Ok(None),
+            NVME_ITER_ERROR => Err(self
+                .controller
+                .fatal_context("failed to iterate nvme namespaces")),
+            invalid => unreachable!(
+                "invalid nvme controller iteration state ({invalid})",
+            ),
+        }
+    }
+}
+
+impl<'a> Iterator for NamespaceDiscovery<'a> {
+    type Item = Result<Namespace<'a>, NvmeError>;
+
+    fn next(&mut self) -> Option<Result<Namespace<'a>, NvmeError>> {
+        self.internal_step().transpose()
+    }
+}
+
+pub struct Namespace<'a> {
+    inner: *mut nvme_ns_t,
+    controller: &'a Controller<'a>,
+}
+
+impl<'a> Drop for Namespace<'a> {
+    fn drop(&mut self) {
+        unsafe { nvme_ns_fini(self.inner) }
+    }
+}
+
+impl<'a> Namespace<'a> {
+    pub fn get_info(&self) -> Result<NamespaceInfo, NvmeError> {
+        let mut nvme_ns_info: *mut nvme_ns_info_t = std::ptr::null_mut();
+        match unsafe { nvme_ns_info_snap(self.inner, &mut nvme_ns_info) } {
+            true => Ok(unsafe { NamespaceInfo::from_raw(nvme_ns_info) }),
+            false => Err(self
+                .controller
+                .fatal_context("failed to get ns info snapshot")),
+        }
+    }
+
+    pub fn blkdev_attach(&self) -> Result<(), NvmeError> {
+        match unsafe { nvme_ns_bd_attach(self.inner) } {
+            true => Ok(()),
+            false => Err(self
+                .controller
+                .fatal_context("failed to attach blkdev to namespace")),
+        }
+    }
+
+    pub fn blkdev_detach(&self) -> Result<(), NvmeError> {
+        match unsafe { nvme_ns_bd_detach(self.inner) } {
+            true => Ok(()),
+            false => Err(self
+                .controller
+                .fatal_context("failed to detach blkdev to namespace")),
+        }
+    }
+}
+
+pub struct NamespaceInfo(*mut nvme_ns_info_t);
```

Comment:
I think this is right but just double checking - it's okay for this to be alive after all possible parents (namespace, controller, and even handle) have been freed?

Reply:
I confirmed with rm over chat that the ns and ctrl info structures are designed to have independent lifetimes.

Reply:
That being said, maybe I should go back and relax that requirement on `ControllerInfo`?

Reply:
Done.